backend/modules/brain/knowledge_brain_qa.py

```diff
@@ -245,109 +345,268 @@ def calculate_pricing(self):
-    ) -> AsyncIterable:
-        conversational_qa_chain = self.knowledge_qa.get_chain()
-        transformed_history, streamed_chat_history = (
-            self.initialize_streamed_chat_history(chat_id, question)
+            chat_id=chat_id,
+            question=question,
+            answer=answer,
+            metadata=metadata,
+            save_answer=save_answer
         )
-        response_tokens = ""
-        sources = []
-        citations = []
-        first = True
-        async for chunk in conversational_qa_chain.astream(
-            {
-                "question": question.question,
-                "chat_history": transformed_history,
-                "custom_personality": (
-                    self.prompt_to_use.content if self.prompt_to_use else None
-                ),
-            }
-        ):
-            if not streamed_chat_history.metadata:
-                streamed_chat_history.metadata = {}
-            if self.model_compatible_with_function_calling(model=self.model):
-                if chunk.get("answer"):
-                    if first:
-                        gathered = chunk["answer"]
-                        first = False
-                    else:
-                        gathered = gathered + chunk["answer"]
-                        if (
-                            gathered.tool_calls
-                            and gathered.tool_calls[-1].get("args")
-                            and "answer" in gathered.tool_calls[-1]["args"]
-                        ):
-                            # Only send the difference between answer and response_tokens which was the previous answer
-                            answer = gathered.tool_calls[-1]["args"]["answer"]
-                            difference = answer[len(response_tokens) :]
-                            streamed_chat_history.assistant = difference
-                            response_tokens = answer
-
-                            yield f"data: {json.dumps(streamed_chat_history.dict())}"
-                        if (
-                            gathered.tool_calls
-                            and gathered.tool_calls[-1].get("args")
-                            and "citations" in gathered.tool_calls[-1]["args"]
-                        ):
-                            citations = gathered.tool_calls[-1]["args"]["citations"]
-            else:
-                if chunk.get("answer"):
-                    response_tokens += chunk["answer"].content
-                    streamed_chat_history.assistant = chunk["answer"].content
-                    yield f"data: {json.dumps(streamed_chat_history.dict())}"
-
-            if chunk.get("docs"):
-                sources = chunk["docs"]
-
-        sources_list = generate_source(sources, self.brain_id, citations)
-
-        # Serialize the sources list
+    
+    async def generate_stream(self, chat_id: UUID, question: ChatQuestion, save_answer: bool = True) -> AsyncIterable:
```

Comment:
You might have an issue in the way you Yield the result. I don't think it is an async iterable the output.

Reply:
Ok thanks I'll take a look

Reply:
@StanGirard I just pushed a new commit

Reply:
<img width="1107" alt="image" src="https://github.com/QuivrHQ/quivr/assets/19614572/003bdb90-c310-4d6b-9e52-46d06bd8d6a0">


I still get this error :/ 

Don't worry I'll take a look my self, I don't want you to spend to much time on it for now ;)

Reply:
And the issue with the rate limit is probably because you didn't put your credit card on OpenAI for the api 😅

Reply:
> And the issue with the rate limit is probably because you didn't put your credit card on OpenAI for the api 😅

I did, and even bought some credits to get higher RPM, but still got the same error… 

I wanted to try with Ollama, but couldn’t figure out how to make it work either…

Reply:
I foudn the issue :) 

You are not yielding the results. I just had the same bug on another feature ;)

Reply:
Ah! Cool tnx!